ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -268,7 +268,7 @@ densityDisconnect (GenesisWindow sgen) (SecurityParam k) states candidateSuffixe
       -- If the density is 0, the peer should be disconnected. This affects
       -- ChainSync jumping, where genesis windows with no headers prevent jumps
       -- from happening.
-      if ub0 == 0 && not idling0 then pure peer0 else do
+      if ub0 == 0 && hasBlockAfter0 then pure peer0 else do
```

Comment:
From the commit message:

> This is not the same as not idling, since we want to disconnect when the peer is idling and has a header after sgen (if this is ever possible).

To double-check: In case we can forecast further than the Genesis window (which can happen ATM as the forecasting range starts at the intersection between the candidate fragment and the current selection), the peer might already have a block after the Genesis window *and* be idle at the same time. In that case, we want to disconnect when the density is exactly zero :+1:

Reply:
That's my thinking. This code is further refined in a later commit.